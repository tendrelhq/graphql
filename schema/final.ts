import { resolvers, typeDefs } from "@/schema";
import { makeExecutableSchema, mergeSchemas } from "@graphql-tools/schema";
import { getSchema } from "./v1.schema";

const legacySchema = makeExecutableSchema({ typeDefs, resolvers });
const gratsSchema = getSchema();

export const schema = mergeSchemas({ schemas: [legacySchema, gratsSchema] });
