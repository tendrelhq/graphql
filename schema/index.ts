export type * from "./types";
export type * from "./__generated__/types.generated";
export * from "./__generated__/typeDefs.generated";
export * from "./__generated__/resolvers.generated";
