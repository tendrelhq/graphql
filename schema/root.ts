/** @gqlType */
export type Query = unknown;

/** @gqlType */
export type Mutation = unknown;
