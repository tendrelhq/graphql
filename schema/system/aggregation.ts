/** @gqlType */
export type Aggregate = {
  /**
   * The group, or bucket, that uniquely identifies this aggregate.
   * For example, this will be one of the `groupByTag`s passed to `trackingAgg`.
   *
   * @gqlField
   */
  group: string;

  /**
   * The computed aggregate value.
   *
   * Currently, this will always be a string value representing a duration in
   * seconds, e.g. "360" -> 360 seconds. `null` will be returned when no such
   * aggregate can be computed, e.g. "time in planned downtime" when no "planned
   * downtime" events exist.
   *
   * @gqlField
   */
  value: string | null;
};
