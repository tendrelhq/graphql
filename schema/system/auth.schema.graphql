interface Identity implements Component {
  id: Entity!
}
