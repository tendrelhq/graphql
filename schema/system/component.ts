import { sql } from "@/datasources/postgres";
import type { ID, Int } from "grats";
import type { Fragment } from "postgres";
import { match } from "ts-pattern";
import type { Context } from "../types";
import { DisplayName } from "./component/name";
import type { Task } from "./component/task";
import type { Connection } from "./pagination";
import type { Timestamp } from "./temporal";

/**
 * Components characterize Entities as possessing a particular trait.
 * They are just simple structs, holding all data necessary to model that trait.
 *
 * @gqlInterface
 */
export interface Component {
  readonly __typename: string;

  /**
   * @gqlField
   * @killsParentOnException
   */
  readonly id: ID;
}

export const field$fragment: Fragment = sql`
select
    f._name,
    f.id,
    case
        when f.type = 'Boolean'
        then
            jsonb_build_object(
                '__typename', 'BooleanValue', 'boolean', f.value::boolean
            )
        when f.type = 'Date'
        then
            jsonb_build_object(
                '__typename', 'TimestampValue', 'timestamp', f.value::timestamptz
            )
        when f.type = 'Number'
        then
            jsonb_build_object(
                '__typename', 'NumberValue', 'number', f.value::numeric
            )
        when f.type = 'String'
        then
            jsonb_build_object(
                '__typename', 'StringValue', 'string', f.value
            )
        else '{}'::jsonb
    end as value,
    case
        when f.type = 'Boolean' then 'boolean'
        when f.type = 'Date' then 'timestamp'
        when f.type = 'Number' then 'number'
        when f.type = 'String' then 'string'
        else 'unknown'
    end as "valueType"
from field as f
`;

/**
 * TODO: description.
 *
 * @gqlField
 */
export async function fields(
  parent: Task,
  _ctx: Context,
): Promise<Connection<Field>> {
  if (parent._type !== "workinstance" && parent._type !== "worktemplate") {
    console.warn(`Underlying type '${parent._type}' does not support fields`);
    return {
      edges: [],
      pageInfo: {
        hasNextPage: false,
        hasPreviousPage: false,
      },
      totalCount: 0,
    };
  }

  const rows = await match(parent._type)
    .with(
      "workinstance",
      () => sql<Field[]>`
        with
            field as (
                select
                    encode(
                        ('workresultinstance:' || wi.id || ':' || wr.id)::bytea, 'base64'
                    ) as id,
                    encode(('name:' || n.languagemasteruuid)::bytea, 'base64') as _name,
                    wi.workinstanceid as _id,
                    wr.workresultid as _field,
                    t.systagtype as type,
                    wri.workresultinstancevalue as value
                from public.workinstance as wi
                inner join
                    public.workresultinstance as wri
                    on wi.workinstanceid = wri.workresultinstanceworkinstanceid
                inner join
                    public.workresult as wr
                    on wri.workresultinstanceworkresultid = wr.workresultid
                    and (
                        wr.workresultisprimary = false
                        or (
                            wr.workresultisprimary = true
                            and wr.workresultentitytypeid is null
                            and wr.workresulttypeid != 737  -- Time At Task :heavy-sigh:
                        )
                    )
                inner join
                    public.languagemaster as n
                    on wr.workresultlanguagemasterid = n.languagemasterid
                inner join public.systag as t on wr.workresulttypeid = t.systagid
                where wi.id = ${parent._id}
                order by wr.workresultorder asc, wr.workresultid asc
            )
        ${field$fragment}
      `,
    )
    .with(
      "worktemplate",
      () => sql<Field[]>`
        with
            field as (
                select
                    encode(('workresult:' || wr.id)::bytea, 'base64') as id,
                    encode(('name:' || n.languagemasteruuid)::bytea, 'base64') as "_name",
                    wr.workresultid as _field,
                    t.systagtype as type,
                    wr.workresultdefaultvalue as value
                from public.worktemplate as wt
                inner join
                    public.workresult as wr
                    on wt.worktemplateid = wr.workresultworktemplateid
                    and (wr.workresultenddate is null or wr.workresultenddate > now())
                    and (
                        wr.workresultisprimary = false
                        or (
                            wr.workresultisprimary = true
                            and wr.workresultentitytypeid is null
                            and wr.workresulttypeid != 737  -- Time At Task :heavy-sigh:
                        )
                    )
                inner join
                    public.languagemaster as n
                    on wr.workresultlanguagemasterid = n.languagemasterid
                inner join public.systag as t on wr.workresulttypeid = t.systagid
                where wt.id = ${parent._id}
                order by wr.workresultorder asc, wr.workresultid asc
            )
        ${field$fragment}
      `,
    )
    //
    .otherwise((_: never) => []);

  return {
    edges: rows.map(row => ({ cursor: row.id, node: row })),
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    totalCount: rows.length,
  };
}

/** @gqlType */
export type Field = {
  _name: ID;

  /**
   * Unique identifier for this Field.
   *
   * @gqlField
   */
  id: ID;

  /**
   * The value for this Field, if any. This field will always be present (when
   * requested) for the given Field so as to convey the underlying data type of
   * the (raw data) value. The underlying (raw data) value can be `null`.
   *
   * @gqlField
   */
  value: Value;

  /**
   * The type of data underlying `value`. This is provided as a convenience when
   * interacting with field-level edits through other apis.
   *
   * @gqlField
   */
  valueType: ValueType;
};

/**
 * Display name for a Field.
 *
 * @gqlField
 */
export function name(field: Field): DisplayName {
  return new DisplayName(field._name);
}

/** @gqlInput */
export type FieldInput = {
  field: ID;
  value?: ValueInput | null;
  /**
   * Must match the type of the `value`, e.g.:
   * ```typescript
   * if (field.valueType === "string") {
   *   assert("string" in field.value);
   * }
   * ```
   */
  valueType: ValueType;
};

/** @gqlEnum */
export type ValueType =
  | "boolean"
  | "entity"
  | "number"
  | "string"
  | "timestamp"
  /** For backwards compatibility. */
  | "unknown";

/** @gqlUnion */
export type Value =
  | BooleanValue
  | EntityValue
  | NumberValue
  | StringValue
  | TimestampValue;

/** @gqlType */
type BooleanValue = {
  __typename: "BooleanValue";

  /** @gqlField */
  boolean: boolean | null;
};

/** @gqlType */
type EntityValue = {
  __typename: "EntityValue";

  /** @gqlField */
  entity: Component | null;
};

/** @gqlType */
type NumberValue = {
  __typename: "NumberValue";

  /** @gqlField */
  number: Int | null;
};

/** @gqlType */
type StringValue = {
  __typename: "StringValue";

  /** @gqlField */
  string: string | null;
};

/** @gqlType */
type TimestampValue = {
  __typename: "TimestampValue";

  /** @gqlField */
  timestamp: Timestamp | null;
};

/**
 * @gqlInput
 * @oneOf
 */
export type ValueInput =
  // Boolean
  | {
      boolean: boolean;
    }
  // Entity
  | {
      id: ID;
    }
  // Number
  | {
      number: Int;
    }
  // String
  | {
      string: string;
    }
  // Temporal
  | {
      /**
       * Date in either ISO or epoch millisecond format.
       */
      timestamp: string;
    };
