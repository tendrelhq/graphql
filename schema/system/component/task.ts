import { sql } from "@/datasources/postgres";
import type { Trackable } from "@/schema/platform/tracking";
import type { Mutation } from "@/schema/root";
import type { Component } from "@/schema/system/component";
import type { Overridable } from "@/schema/system/overridable";
import type { Timestamp } from "@/schema/system/temporal";
import type { Context } from "@/schema/types";
import { assertNonNull } from "@/util";
import { GraphQLError } from "graphql";
import type { ID } from "grats";
import { match } from "ts-pattern";
import { decodeGlobalId } from "..";
import type { Refetchable } from "../node";
import type { Connection } from "../pagination";
import type { DisplayName, Named } from "./name";

export type ConstructorArgs = {
  id: ID;
};

/**
 * A system-level component that identifies an Entity as being applicable to
 * tendrel's internal "task processing pipeline". In practice, Tasks most often
 * represent "jobs" performed by humans. However, this need not always be the
 * case.
 *
 * Technically speaking, a Task represents a (1) *named asynchronous process*
 * that (2) exists in one of three states: open, in progress, or closed.
 *
 * @gqlType
 */
export class Task implements Component, Named, Refetchable, Trackable {
  readonly __typename = "Task" as const;
  readonly _type: string;
  readonly _id: string;
  readonly id: ID;

  constructor(
    args: ConstructorArgs,
    private ctx: Context,
  ) {
    // Note that Postgres will sometimes add newlines when we `encode(...)`.
    this.id = args.id.replace(/\n/g, "");
    // Private.
    const { type, id } = decodeGlobalId(this.id);
    this._type = type;
    this._id = id;
  }

  /** @gqlField */
  async displayName(): Promise<DisplayName> {
    return await this.ctx.orm.displayName.load(this.id);
  }

  async root(): Promise<Task | null> {
    if (this._type !== "workinstance") return null;

    const [row] = await sql<[{ id: ID }?]>`
      SELECT encode(('workinstance:' || og.id)::bytea, 'base64') AS id
      FROM public.workinstance AS wi
      INNER JOIN public.workinstance AS og
          ON wi.workinstanceoriginatorid = og.workinstanceid
      WHERE
          wi.id = ${this._id}
          AND wi.workinstanceid IS DISTINCT FROM wi.workinstanceoriginatorid;
    `;

    // If there is no row, there is no originator. This implies that this Task
    // is the root of the chain, i.e. the originator.
    if (!row) return this;

    return new Task({ id: row.id }, this.ctx);
  }

  /** @gqlField */
  async state(): Promise<TaskState | null> {
    // Only workinstances have statuses.
    if (this._type !== "workinstance") return null;

    const [row] = await sql<
      [
        {
          status: string;
          create_date: string;
          target_start_date?: string;
          ov_target_start_date?: string;
          start_date?: string;
          ov_start_date?: string;
          close_date?: string;
          ov_close_date?: string;
          time_zone: string;
        }?,
      ]
    >`
      SELECT
          wis.systagtype AS status,
          wi.workinstancecreateddate AS create_date,
          wi.workinstancetargetstartdate AS target_start_date,
          nullif(ov_target.workresultinstancevalue, '')::timestamptz AS ov_target_start_date,
          wi.workinstancestartdate AS start_date,
          nullif(ov_start.workresultinstancevalue, '')::timestamptz AS ov_start_date,
          wi.workinstancecompleteddate AS close_date,
          nullif(ov_close.workresultinstancevalue, '')::timestamptz AS ov_close_date,
          wi.workinstancetimezone AS time_zone
      FROM public.workinstance AS wi
      INNER JOIN public.systag AS wis
          ON wi.workinstancestatusid = wis.systagid
      LEFT JOIN public.workresult AS ov_target_f
          ON
              wi.workinstanceworktemplateid = ov_target_f.workresultworktemplateid
              AND ov_target_f.workresultisprimary = true
              AND ov_target_f.workresulttypeid IN (
                  SELECT systagid
                  FROM public.systag
                  WHERE systagtype = 'Override Target Start Time'
              )
      LEFT JOIN public.workresultinstance AS ov_target
          ON
              wi.workinstanceid = ov_target.workresultinstanceworkinstanceid
              AND ov_target_f.workresultid = ov_target.workresultinstanceworkresultid
      LEFT JOIN public.workresult AS ov_start_f
          ON
              wi.workinstanceworktemplateid = ov_start_f.workresultworktemplateid
              AND ov_start_f.workresultisprimary = true
              AND ov_start_f.workresulttypeid IN (
                  SELECT systagid
                  FROM public.systag
                  WHERE systagtype = 'Override Start Time'
              )
      LEFT JOIN public.workresultinstance AS ov_start
          ON
              wi.workinstanceid = ov_start.workresultinstanceworkinstanceid
              AND ov_start_f.workresultid = ov_start.workresultinstanceworkresultid
      LEFT JOIN public.workresult AS ov_close_f
          ON
              wi.workinstanceworktemplateid = ov_close_f.workresultworktemplateid
              AND ov_close_f.workresultisprimary = true
              AND ov_close_f.workresulttypeid IN (
                  SELECT systagid
                  FROM public.systag
                  WHERE systagtype = 'Override Start Time'
              )
      LEFT JOIN public.workresultinstance AS ov_close
          ON
              wi.workinstanceid = ov_close.workresultinstanceworkinstanceid
              AND ov_close_f.workresultid = ov_close.workresultinstanceworkresultid
      WHERE wi.id = ${this._id};
    `;

    if (!row) {
      console.warn(`No TaskState for ${this}... which is odd?`);
      return null;
    }

    return match(row.status)
      .with(
        "Open",
        () =>
          ({
            __typename: "Open" as const,
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
          }) satisfies Open,
      )
      .with(
        "In Progress",
        () =>
          ({
            __typename: "InProgress",
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
            inProgressAt: {
              override:
                row.start_date && row.ov_start_date
                  ? {
                      previousValue: {
                        epochMilliseconds: row.start_date,
                        timeZone: row.time_zone,
                      },
                    }
                  : null,
              value: {
                epochMilliseconds: assertNonNull(
                  row.ov_start_date ?? row.start_date,
                  `Task ${this}, in state 'In Progress', has no start date`,
                ),
                timeZone: row.time_zone,
              },
            },
          }) satisfies InProgress,
      )
      .with(
        "Cancelled",
        () =>
          ({
            __typename: "Closed",
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
            inProgressAt: row.start_date
              ? {
                  override: row.ov_start_date
                    ? {
                        previousValue: {
                          epochMilliseconds: row.start_date,
                          timeZone: row.time_zone,
                        },
                      }
                    : null,
                  value: {
                    epochMilliseconds: row.ov_start_date ?? row.start_date,
                    timeZone: row.time_zone,
                  },
                }
              : null,
            closedAt: {
              override:
                row.close_date && row.ov_close_date
                  ? {
                      previousValue: {
                        epochMilliseconds: row.close_date,
                        timeZone: row.time_zone,
                      },
                    }
                  : null,
              value: {
                epochMilliseconds: assertNonNull(
                  row.ov_close_date ?? row.close_date,
                  `Task ${this}, in state 'Cancelled', has no close date`,
                ),
                timeZone: row.time_zone,
              },
            },
          }) satisfies Closed,
      )
      .with(
        "Complete",
        () =>
          ({
            __typename: "Closed",
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
            inProgressAt: row.start_date
              ? {
                  override: row.ov_start_date
                    ? {
                        previousValue: {
                          epochMilliseconds: row.start_date,
                          timeZone: row.time_zone,
                        },
                      }
                    : null,
                  value: {
                    epochMilliseconds: row.ov_start_date ?? row.start_date,
                    timeZone: row.time_zone,
                  },
                }
              : null,
            closedAt: {
              override:
                row.close_date && row.ov_close_date
                  ? {
                      previousValue: {
                        epochMilliseconds: row.close_date,
                        timeZone: row.time_zone,
                      },
                    }
                  : null,
              value: {
                epochMilliseconds: assertNonNull(
                  row.ov_close_date ?? row.close_date,
                  `Task ${this}, in state 'Closed', has no close date`,
                ),
                timeZone: row.time_zone,
              },
            },
          }) satisfies Closed,
      )
      .otherwise(s => {
        console.warn(`Unknown TaskState type '${s}'`);
        return null;
      });
  }

  /**
   * Entrypoint into the "tracking system(s)" for the given Task.
   * At the moment, sub-task tracking is not supported and therefore `null` will
   * always be returned for this field.
   *
   * @gqlField
   */
  async tracking(): Promise<Connection<Trackable> | null> {
    return null;
  }

  toString() {
    return `'${this.id}' (${this._type}:${this._id})`;
  }
}

// FIXME: this might be a confusing name in combination with TaskStateMachine...
// Probably should just name it TaskStatus?
/** @gqlUnion */
export type TaskState = Open | InProgress | Closed;

/** @gqlType */
export type Open = {
  __typename: "Open";

  /** @gqlField */
  dueAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  openedAt: Overridable<Timestamp>;
  /** @gqlField */
  openedBy?: string;
};

/** @gqlType */
export type InProgress = {
  __typename: "InProgress";

  /** @gqlField */
  dueAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  openedAt: Overridable<Timestamp>;
  /** @gqlField */
  openedBy?: string | null;
  /** @gqlField */
  inProgressAt: Overridable<Timestamp>;
  /** @gqlField */
  inProgressBy?: string | null;
};

/** @gqlType */
export type Closed = {
  __typename: "Closed";

  /** @gqlField */
  dueAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  openedAt: Overridable<Timestamp>;
  /** @gqlField */
  openedBy?: string | null;
  /** @gqlField */
  inProgressAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  inProgressBy?: string | null;
  /** @gqlField */
  closedAt: Overridable<Timestamp>;
  /** @gqlField */
  closedBecause?: string | null;
  /** @gqlField */
  closedBy?: string;
};

/**
 * Similar to task_fsm's advance method, this method advances a Task through its
 * internal state machine. A Task's state machine has its finite set defined by
 * the variants of the {@link TaskState} union, with a dag of the form:
 * ```
 * Open -> InProgress -> Closed
 * ```
 */
export async function advance(t: Task): Promise<Task> {
  if (t._type !== "workinstance") {
    // Punt on this for now. We can come back to it.
    // This is, at least at present, used solely by the `advance` implementation
    // for StateMachine<Task>.
    throw "not yet implemented - lazy instantiation of a Task";
  }

  // As before, we assume this is the "stop task" flow.
  await sql`
    UPDATE public.workinstance
    SET
        workinstancestatusid = 710,
        workinstancecompleteddate = now(),
        workinstancemodifieddate = now()
        -- TODO: workinstancemodifiedby
    WHERE id = ${t._id};
  `;

  return t;
}
