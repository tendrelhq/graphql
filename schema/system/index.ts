import z, { type Infer } from "myzod";

const GlobalId = z.string().map(id => {
  const parts = id.split(":");
  if (parts.length !== 2) {
    throw "invariant violated: invalid global identifier";
  }
  return {
    type: parts[0],
    id: parts[1],
  };
});

/**
 * Global Identifier. Externally it just an opaque string. Internally it may
 * have some meaning. Right now this is just the base64 encoding of the
 * object's underlying type (e.g. workinstance) and its uuid (i.e. primary
 * key).
 */
type GlobalId = Infer<typeof GlobalId>;

export function decodeGlobalId(id: unknown): GlobalId {
  if (typeof id !== "string") {
    throw "invariant violated: global ids should be string";
  }
  return GlobalId.parse(
    Buffer.from(decodeURIComponent(id), "base64").toString(),
  );
}

export function encodeGlobalId({ type, id }: GlobalId) {
  return Buffer.from(`${type}:${id}`).toString("base64");
}

// biome-ignore lint/suspicious/noExplicitAny:
export function isGlobalId(value: any): value is GlobalId {
  if (typeof value?.type === "string" && typeof value?.id === "string") {
    return true;
  }

  return false;
}
