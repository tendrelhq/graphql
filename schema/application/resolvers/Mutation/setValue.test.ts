import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { sql } from "@/datasources/postgres";
import { schema } from "@/schema/final";
import { decodeGlobalId, encodeGlobalId } from "@/schema/system";
import { Task } from "@/schema/system/component/task";
import {
  createTestContext,
  execute,
  findAndEncode,
  getFieldByName,
} from "@/test/prelude";
import { assertNonNull, map } from "@/util";
import { TestSetValueDocument } from "./setValue.test.generated";

const ctx = await createTestContext();

describe("setValue", () => {
  let CUSTOMER: string;
  let TASK: Task;

  test("set", async () => {
    const field = await getFieldByName(TASK, "Run Output");
    const result = await execute(schema, TestSetValueDocument, {
      parent: TASK.id,
      entity: field,
      input: {
        number: {
          value: 42,
        },
      },
    });

    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("set: dynamic content", async () => {
    const field = await getFieldByName(TASK, "Comments");
    const result = await execute(schema, TestSetValueDocument, {
      parent: TASK.id,
      entity: field,
      input: {
        string: {
          value: "Hello world",
        },
      },
    });

    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();

    // TODO: move to prelude
    const { id, suffix } = decodeGlobalId(field);
    const [row] = await sql`
      select
          languagemastersource as content,
          languagemastersourcelanguagetypeid as language
      from public.workresultinstance
      inner join
          public.languagemaster
          on workresultinstancevaluelanguagemasterid = languagemasterid
      where
          workresultinstanceworkinstanceid = (
              select workinstanceid
              from public.workinstance
              where id = ${id}
          )
          and workresultinstanceworkresultid = (
              select workresultid
              from public.workresult
              where id = ${assertNonNull(suffix?.at(0))}
          )
    `;
    expect(row).toEqual({
      content: "Hello world",
      language: 20n,
    });
  });

  test("entity is not mutable", async () => {
    const result = await execute(schema, TestSetValueDocument, {
      parent: TASK.id,
      entity: encodeGlobalId({
        type: "foo",
        id: "foo",
      }),
      input: {
        string: {
          value: "hello world",
        },
      },
    });
    expect(result).toMatchSnapshot();
  });

  test("global id invariant", async () => {
    const result = await execute(schema, TestSetValueDocument, {
      parent: TASK.id,
      entity: encodeGlobalId({
        type: "workresultinstance",
        id: "foo",
      }),
      input: {
        string: {
          value: "hello world",
        },
      },
    });
    expect(result).toMatchSnapshot();
  });

  beforeAll(async () => {
    const logs = await sql<{ op: string; id: string }[]>`
      select *
      from
          runtime.create_demo(
              customer_name := 'Frozen Tendy Factory',
              admins := (
                  select array_agg(workeruuid)
                  from public.worker
                  where workeridentityid = ${ctx.auth.userId}
              ),
              modified_by := 895
          )
      ;
    `;
    CUSTOMER = findAndEncode("customer", "organization", logs);
    TASK = map(
      findAndEncode("instance", "workinstance", logs),
      id => new Task({ id }, ctx),
    );
    console.log(TASK._type, TASK._id);
  });

  afterAll(async () => {
    const { id } = decodeGlobalId(CUSTOMER);
    process.stdout.write("Cleaning up... ");
    const [row] = await sql<[{ ok: string }]>`
      select runtime.destroy_demo(${id}) as ok;
    `;
    console.log(row.ok);
  });
});
