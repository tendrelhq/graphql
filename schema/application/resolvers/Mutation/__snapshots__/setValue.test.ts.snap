// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`setValue set 1`] = `
{
  "setValue": {
    "__typename": "SetValuePayload",
    "delta": 1,
    "node": {
      "__typename": "ChecklistResult",
      "status": {
        "__typename": "ChecklistOpen",
      },
      "widget": {
        "__typename": "NumberWidget",
        "number": 42,
      },
    },
    "parent": {
      "__typename": "Checklist",
      "status": {
        "__typename": "ChecklistOpen",
      },
    },
  },
}
`;

exports[`setValue set: dynamic content 1`] = `
{
  "setValue": {
    "__typename": "SetValuePayload",
    "delta": 1,
    "node": {
      "__typename": "ChecklistResult",
      "status": {
        "__typename": "ChecklistOpen",
      },
      "widget": {
        "__typename": "StringWidget",
        "string": "Hello world",
      },
    },
    "parent": {
      "__typename": "Checklist",
      "status": {
        "__typename": "ChecklistOpen",
      },
    },
  },
}
`;

exports[`setValue entity is not mutable 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Entity is not mutable: foo],
  ],
}
`;

exports[`setValue global id invariant 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Unexpected error value: "invariant violated"],
  ],
}
`;
